### PR TITLE
busybox: install as setuid root for FEATURE_SUID=y

### PR DESCRIPTION
**Pull request by hmh**
A few busybox applets require privileged capabilities to work (such as
"traceroute" and "su").  This restriction is not relevant when
everything runs as root, but THAT is even less optimal than trusting
busybox traceroute, ping, and a few other applets to safely run setuid
to root.

Allow non-root users to use such busybox applets by installing busybox
setuid root when busybox has been compiled with support to drop setuid
root status for applets that don't need it.

Note that BUSYBOX_CONFIG_FEATURE_SUID *IS* enabled by default, so
busybox WILL BE installed setuid root by default with this change.

Full-featured capabilities support would be better (so that, e.g., the
traceroute applet could run unprivileged but with CAP_NET_RAW+ep), but
that requires support from lots of OpenWRT components (kernel,
filesystems and overlay, ROM filesystem image and build system, etc).

Note that we can't just "chmod u+s /bin/busybox" at runtime as a
"cheaper" solution: it would waste approximately 200KiB of FLASH (the
whole /bin/busybox binary gets copied into the overlay).

Signed-off-by: Henrique de Moraes Holschuh <henrique@nic.br>

### CODE REVIEW

**Comment by hmh** [comment]
PS: tested to work on older OpenWRT: we have it backported to 15.05, 18.06 and 19.07.  I assume it would also work on LEDE-17.01.

**Comment by jeffsf** [comment]
Why is this desirable? 

A user that has "normal" configuration of the device is logged in as root.

A sophisticated user that adds non-root accounts can install `sudo`.

**Comment by hmh** [comment]
Think system services (scripts) not running as root.  We have some that, e.g. traceroute relevant targets and report to the user any issues it finds.  We don't want to run the entire thing as root if we can help it...

**Comment by jeffsf** [comment]
Let me rephrase then, "Why is this desirable for _all_ users?" as it seems that this is to fulfill needs of custom firmware, over which you have control of the source.

At least for me, I consider any more than the absolute minimum number of SUID binaries a security risk. Something as complex as busybox seems especially risky.

**Comment by hauke** [comment]
Will only ping and so on be executed as root with these changes or every busybox command?

**Comment by hmh** [comment]
@hauke:
Busybox dispatcher annotates which applets may run with setuid privileges, all other applets will have the setuid privilege dropped *before* doing any work.  This is handled by the busybox main code (and not individually by each applet).

It will affect su, wall, crontab, passwd, login, ping, traceroute, mount and findfs.  The code for these applets has been written with SUID privileges in mind.

(this list is not really exaustive, it is true for the version of busybox I inspected.  grep -r SUID_MAYBE and SUID_REQUIRE in the busybox source to detect which applets would run with SUID privileges)

Other applets WILL NOT RUN as setuid root, busybox will DROP the suid privilege prior to running them.

**Comment by hmh** [comment]
@jeffsf:
Anywhere else, I would agree with you. But this is OpenWRT, with its lack of full support for Linux capabilities (as far as I know -- would love to be proven wrong).

Having select code that has been engineered to run with setuid root privileges running setuid root is a way to decrease the amount of code that has to run as root, when one cannot use more specific capabilities to do so.

Anyway, if there is a need for a new config option to actually have busybox work as it should (ping, traceroute and crontab does not require root to work), I can change the patch to work that way too.

But it seems weird to have BUSYBOX_CONFIG_FEATURE_SUID enabled by default *WITHOUT* the setuid root busybox, since that pretty much means it is there just so that "crontab" and "mount" tell you that they require root to work.  This is why I did not add an extra CONFIG option.

**Comment by hmh** [comment]
Any words on this?

As I said, I can change the PR to add a new knob instead of extending BUSYBOX_CONFIG_FEATURE_SUID if such a change would allow this PR to get merged...

**Comment by ynezz** [comment]
> Any words on this?

Patience.

> But it seems weird to have BUSYBOX_CONFIG_FEATURE_SUID enabled by default WITHOUT the setuid root busybox

Indeed, but to me the proper fix right now would be to disable BUSYBOX_CONFIG_FEATURE_SUID `(default n)` in additional commit, keeping the current status quo. If you would like to recommend having it enabled by default, then it would make more sense to send it as separate patch to the mailing list (once this fixes are merged).

**Comment by hmh** [comment]
@ynezz:
> Indeed, but to me the proper fix right now would be to disable BUSYBOX_CONFIG_FEATURE_SUID `(default n)` in additional commit, keeping the current status quo. If you would like to recommend having it enabled by default, then it would make more sense to send it as separate patch to the mailing list (once this fixes are merged).

Done: added a commit that flips BUSYBOX_DEFAULT_FEATURE_SUID to "n", and build-tested it.

**Comment by ynezz** [comment]
Thanks! Pulled into my staging tree at https://git.openwrt.org/openwrt/staging/ynezz.git

**Comment by ynezz** [comment]
@hmh please check both commits 0e2c127b6efc08b0b7efef4f44ab6a0b88651bd2 and 069c46d06c381f2e0ed77a7d7fb4cfa80f4eacb2 as I've adjusted the commits little bit, so they make more sense, added missing SoB and replaced `chmod` with `INSTALL_SUID`. Please let me know if you're fine with those changes so I can push them, thanks.

**Comment by hmh** [comment]
@ynezz: the commits look fine.

I used "chmod u+s" in the original commit because I don't like to do INSTALL_* on top of something that was already subject to INSTALL_*  (in this case, /bin/busybox comes first from a INSTALL_DIR and then from a INSTALL_SUID on top) for cosmetic reasons, but it *should* work just fine, so that's beside the point.

Thanks!

**Comment by hmh** [comment]
BTW, please consider back-porting to 18.06 as well, since it doesn't change the status-quo ;-)
